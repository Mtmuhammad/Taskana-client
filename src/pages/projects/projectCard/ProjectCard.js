// Component that displays project information on project page

import { useState, useEffect } from "react";
import useAxiosPrivate from "../../../hooks/useAxiosPrivate";
import { useNavigate, useLocation, Link } from "react-router-dom";
import useAuth from "../../../hooks/useAuth";
import "./ProjectCard.scss";
import DeleteModal from "../deleteProject/DeleteModal";
import EditModal from "../editProject/EditModal";

const ProjectCard = ({
  setProjects,
  projects,
  project,
  setSuccess,
  setErrMsg,
}) => {
  const { auth } = useAuth();
  const navigate = useNavigate();
  const axiosPrivate = useAxiosPrivate();
  const location = useLocation();

  const [tickets, setTickets] = useState();
  const [info, setInfo] = useState();
  const [closedTickets, setClosedTickets] = useState();
  const [progress, setProgress] = useState(0);

  // get all Tickets
  useEffect(() => {
    let isMounted = true;
    const controller = new AbortController();

    const getTickets = async () => {
      try {
        const res = await axiosPrivate.get(`/tickets`, {
          signal: controller.signal,
        });
        const result = await axiosPrivate.get(`/projects/${project.id}`, {
          signal: controller.signal,
        });

        setInfo(result.data.project);
        const tickets = res.data.tickets.filter(
          (ticket) => ticket.projectId === project.id
        );
        const closed = res.data.tickets.filter(
          (ticket) => ticket.status === "Completed"
        );
        isMounted && setTickets(tickets);
        setClosedTickets(closed);

        tickets?.length !== 0 &&
          setProgress(
            Math.round((closedTickets?.length / tickets?.length) * 100) ||
              Number(0)
          );
      } catch (err) {
        console.log(err);
        navigate("/login", { state: { from: location }, replace: true });
      }
    };

    getTickets();

    return () => {
      isMounted = false;
      controller.abort();
    };
  }, [closedTickets?.length, project?.id, axiosPrivate, location, navigate]);

  return (
    <div className="col-lg-4 col-md-6">
      <div data-testid="project-card" className="mb-5 project-card card">
        <div className="card-body">
          <div className="d-flex align-items-center justify-content-between">
            <div className="project-name">
              <h4 data-testid="project-header" className="h3 mb-0 fw-bold mb-2">
                {project?.name}
              </h4>
            </div>
            {auth?.user?.isAdmin && (
              <div
                className="btn-group"
                role="group"
                aria-label="Basic outlined example"
              >
                <button
                  type="button"
                  className="btn btn-outline-secondary"
                  data-bs-toggle="modal"
                  data-bs-target={`#editProject${project?.id}`}
                >
                  <i className="bx bxs-edit"></i>
                </button>
                <button
                  type="button"
                  className="btn btn-outline-secondary"
                  data-bs-toggle="modal"
                  data-bs-target={`#deleteProject${project?.id}`}
                >
                  <i className="bx bx-trash"></i>
                </button>
              </div>
            )}
          </div>

          <div className="row g-2 pt-4">
            <div className="col-6">
              <div className="d-flex align-items-center">
                <i
                  style={{ background: "#15cfff90" }}
                  className="rounded-circle p-2 bx bx-paperclip"
                  data-testid="project-card-icon"
                ></i>
                <span data-testid="project-date" className="ms-2">
                  Created: {project?.date}{" "}
                </span>
              </div>
            </div>
            <div className="col-6">
              <div className="d-flex align-items-center">
                <i
                  style={{ background: "#00d09c90" }}
                  className="rounded-circle p-2 bx bxs-hourglass-top"
                  data-testid="project-card-icon"
                ></i>
                <span data-testid="project-deadline" className="ms-2">
                  Deadline: {project?.deadline}{" "}
                </span>
              </div>
            </div>
            <div className="col-6">
              <div className="d-flex align-items-center">
                <i
                  style={{ background: "#ffc36d90" }}
                  className="rounded-circle p-2 bx bxs-report"
                  data-testid="project-card-icon"
                ></i>
                <span data-testid="project-status" className="ms-2">
                  Status: {project?.status}
                </span>
              </div>
            </div>
            <div className="col-6">
              <div className="d-flex align-items-center">
                <i
                  style={{ background: "#ff505090" }}
                  className="rounded-circle p-2 bx bx-note"
                  data-testid="project-card-icon"
                ></i>
                <span className="ms-2" data-testid="project-tickets">
                  {tickets && tickets.length} tickets
                </span>
              </div>
            </div>
          </div>
          <div className="dividers-block"></div>
          <div data-testid="project-description" className="mt-4 h5">{info?.description}</div>
          <div className="d-flex align-items-center justify-content-center mt-3 mb-2">
            <h5 className="fw-bold mb-0">Progress</h5>
          </div>
          <div
            style={{ height: "2rem" }}
            data-testid="project-progress"
            className="progress mt-3"
          >
            {progress >= 0 && (
              <div
                className="progress-bar bg-warning"
                role="progressbar"
                style={{
                  padding: "1rem 0",
                  fontSize: "1.2rem",
                  width: `${progress}%`,
                }}
                aria-valuenow={progress}
                aria-valuemin="0"
                aria-valuemax="100"
              >
                {progress}%
              </div>
            )}
          </div>
        </div>
      </div>

      <DeleteModal
        setProjects={setProjects}
        projects={projects}
        setErrMsg={setErrMsg}
        setSuccess={setSuccess}
        project={project}
      />
      <EditModal
        setProjects={setProjects}
        projects={projects}
        setErrMsg={setErrMsg}
        setSuccess={setSuccess}
        project={project}
      />
    </div>
  );
};

export default ProjectCard;
