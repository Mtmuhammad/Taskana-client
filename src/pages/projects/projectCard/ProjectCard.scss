@use "../../../styles/abstracts/_variables" as *;
@use "../../../styles/abstracts/_mixins" as *;

.project-card {
   background-color: var(--primary-color-light);
   color: var(--text-color);

   & .card-body {
      padding: 2rem;
   }

   & i {
      font-size: 1.5rem;
      margin-top: 0.4rem;

      &.bxs-edit {
         color: $primary-gold;
      }
      &.bx-trash {
         color: $primary-red;
      }
   }

   & span {
      font-size: 1.3rem;
   }

   & h4 {
      font-size: 1.6rem;
   }

   
}

.create-btn {
   background-color: $primary-green;
   color: var(--text-color);
   font-size: 1.5rem;
   padding: 0.75rem 1rem;

   &:hover {
      background-color: $primary-green;
   }
}

.modal{
   &-content {
      font-size: 1.5rem;
   }

   & h5 {
      font-size: 1.7rem;
   }

   & input, textarea {
      font-size: 1.2rem;
   }
}

.msg {
   width: 50%;
    margin: 0 auto 2rem;
}
