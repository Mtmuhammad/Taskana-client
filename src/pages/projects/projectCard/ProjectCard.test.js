import React from "react";
import { render, screen } from "@testing-library/react";
import ProjectCard from "./ProjectCard";
import { BrowserRouter } from "react-router-dom";

describe("smoke and snapshot tests", () => {
  test("should render without crashing", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}} />, { wrapper: BrowserRouter });
  });

  test("should match snapshot", () => {
    const { asFragment } = render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("project card", () => {
  test("should render project card", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const card = screen.getByTestId("project-card");
    expect(card).toBeInTheDocument();
    expect(card).toHaveClass("project-card card");
  });
  test("should render project name", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const name = screen.getByTestId("project-header");
    expect(name).toBeInTheDocument();
    expect(name).toHaveClass("h3 fw-bold");
  });
  test("should render project icons", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const icons = screen.getAllByTestId("project-card-icon");
    expect(icons.length).toBe(4);
  });
  test("should render project date", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const date = screen.getByTestId("project-date");
    expect(date).toBeInTheDocument();
    expect(date).toHaveClass("ms-2");
  });
  test("should render project deadline", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const deadline = screen.getByTestId("project-deadline");
    expect(deadline).toBeInTheDocument();
    expect(deadline).toHaveClass("ms-2");
  });
  test("should render project status", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const status = screen.getByTestId("project-status");
    expect(status).toBeInTheDocument();
    expect(status).toHaveClass("ms-2");
  });
  test("should render project tickets", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}}/>, { wrapper: BrowserRouter });
    const tickets = screen.getByTestId("project-tickets");
    expect(tickets).toBeInTheDocument();
    expect(tickets).toHaveClass("ms-2");
  });
  test("should render project description", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}} />, { wrapper: BrowserRouter });
    const description = screen.getByTestId("project-description");
    expect(description).toBeInTheDocument();
    expect(description).toHaveClass("mt-4 h5");
  });
  test("should render project progress", () => {
    render(<ProjectCard project={{deadline: "07-22-22"}} />, { wrapper: BrowserRouter });
    const progress = screen.getByTestId("project-progress");
    expect(progress).toBeInTheDocument();
    expect(progress).toHaveClass("progress");
  });
});
