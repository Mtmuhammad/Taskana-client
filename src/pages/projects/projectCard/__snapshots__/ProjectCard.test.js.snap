// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-lg-4 col-md-6"
  >
    <div
      class="mb-5 project-card card"
      data-testid="project-card"
    >
      <div
        class="card-body"
      >
        <div
          class="d-flex align-items-center justify-content-between"
        >
          <div
            class="project-name"
          >
            <h4
              class="h3 mb-0 fw-bold mb-2"
              data-testid="project-header"
            />
          </div>
        </div>
        <div
          class="row g-2 pt-4"
        >
          <div
            class="col-6"
          >
            <div
              class="d-flex align-items-center"
            >
              <i
                class="rounded-circle p-2 bx bx-paperclip"
                data-testid="project-card-icon"
                style="background: rgba(21, 207, 255, 0.565);"
              />
              <span
                class="ms-2"
                data-testid="project-date"
              >
                Created:  
              </span>
            </div>
          </div>
          <div
            class="col-6"
          >
            <div
              class="d-flex align-items-center"
            >
              <i
                class="rounded-circle p-2 bx bxs-hourglass-top"
                data-testid="project-card-icon"
                style="background: rgba(0, 208, 156, 0.565);"
              />
              <span
                class="ms-2"
                data-testid="project-deadline"
              >
                Deadline:  
              </span>
            </div>
          </div>
          <div
            class="col-6"
          >
            <div
              class="d-flex align-items-center"
            >
              <i
                class="rounded-circle p-2 bx bxs-report"
                data-testid="project-card-icon"
                style="background: rgba(255, 195, 109, 0.565);"
              />
              <span
                class="ms-2"
                data-testid="project-status"
              >
                Status: 
              </span>
            </div>
          </div>
          <div
            class="col-6"
          >
            <div
              class="d-flex align-items-center"
            >
              <i
                class="rounded-circle p-2 bx bx-note"
                data-testid="project-card-icon"
                style="background: rgba(255, 80, 80, 0.565);"
              />
              <span
                class="ms-2"
                data-testid="project-tickets"
              >
                 tickets
              </span>
            </div>
          </div>
        </div>
        <div
          class="dividers-block"
        />
        <div
          class="d-flex align-items-center justify-content-center mt-3 mb-2"
        >
          <h5
            class="fw-bold mb-0"
          >
            Progress
          </h5>
        </div>
        <div
          class="progress mt-3"
          data-testid="project-progress"
          style="height: 2rem;"
        >
          0
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
